### PR TITLE
Vmo 4208 exit config

### PR DESCRIPTION
**Pull request by bzabos**

### CODE REVIEW

File: src/components/interaction-designer/Block.vue [hunk 1]
```diff
@@ -404,7 +404,7 @@ export default {
         : uuid
     },
 
-    visibleExitTag(key, exit) {
+    exitTagText(key, exit) {
       if (!exit.tag && !exit.semantic_label) {
```
Comment (jcbashdown) [inline src/components/interaction-designer/Block.vue:408]: Should be name

File: src/store/flow/block-types/MobilePrimitives_SelectManyResponseBlockStore.ts [hunk 2]
```diff
@@ -34,24 +34,49 @@ export const actions: ActionTree<ICustomFlowState, IRootState> = {
       uuid: await (new IdGeneratorUuidV4()).generate(),
       tag: 'Default',
       label: 'Default',
+      default: true,
+      // test: '',
+    }
+
+    const errorExitProps: Partial<IBlockExit> = {
+      uuid: await (new IdGeneratorUuidV4()).generate(),
+      tag: 'Error',
```
Comment (jcbashdown) [inline src/store/flow/block-types/MobilePrimitives_SelectManyResponseBlockStore.ts:43]: Should be name, label should be removed

File: src/store/flow/block-types/MobilePrimitives_SelectOneResponseBlockStore.ts [hunk 3]
```diff
@@ -68,11 +73,96 @@ export const getters: GetterTree<ICustomFlowState, IRootState> = {
...
+    // reflow exits based on choices
+    await Promise.all(choiceKeys.map(async (choiceKey, i) => {
+      if (exitsForChoices[i] == null) {
+        const uuid = await (new IdGeneratorUuidV4()).generate()
+        const exit = await dispatch('flow/block_createBlockExitWith', {props: {uuid} as IBlockExit}, {root: true})
+        exitsForChoices.push(exit)
+      }
+
+      Object.assign(exitsForChoices[i], {
+        tag: choiceKey,
```
Comment (jcbashdown) [inline src/store/flow/block-types/MobilePrimitives_SelectOneResponseBlockStore.ts:158]: Change to name

File: src/store/flow/block-types/MobilePrimitives_SelectOneResponseBlockStore.ts [hunk 4]
```diff
@@ -119,45 +211,92 @@ export const actions: ActionTree<ICustomFlowState, IRootState> = {
...
-    const defaultExitProps: Partial<IBlockExit> = {
-      uuid: await (new IdGeneratorUuidV4()).generate(),
-      tag: 'Default',
-      label: 'Default',
-      test: '',
-    }
+    const defaultExit = await dispatch('flow/block_createBlockDefaultExitWith', {
+      props: {
+        uuid: await (new IdGeneratorUuidV4()).generate(),
+        tag: 'Default',
```
Comment (jcbashdown) [inline src/store/flow/block-types/MobilePrimitives_SelectOneResponseBlockStore.ts:296]: etc.

File: src/store/flow/block.ts [hunk 5]
```diff
@@ -74,9 +97,35 @@ export const mutations: MutationTree<IFlowsState> = {
...
+      Vue.set(accumulator, pathKey, nextValue)
+      return nextValue
+    }, metadata!)
+  },
 }
 
 export const actions: ActionTree<IFlowsState, IRootState> = {
+  block_popExitsByLabel({ dispatch, commit, state }, { blockId, exitLabel, shouldUseCache = false }: {blockId: string; exitLabel: string; shouldUseCache: boolean}) {
+    const block = findBlockOnActiveFlowWith(blockId, state as unknown as IContext)
+    const newBlockExits = findBlockExitsRef(block, shouldUseCache).filter((item: IBlockExit) => item.label !== exitLabel)
```
Comment (jcbashdown) [inline src/store/flow/block.ts:120]: We no longer have label. I don't know exactly how we're meant to link exits to choices now?

**Review by jcbashdown — CHANGES_REQUESTED** [review]

File: src/assets/messages.json [hunk 6]
```diff
@@ -151,6 +151,9 @@
     "adds-subscribers-to-the": "Adds contacts to the",
     "admin-csv-file": "Master CSV File",
     "advanced": "Advanced",
```
Comment (safydy-r) [inline src/assets/messages.json:153]: Missing localization here
Screen Shot 2021-07-22 at 10 53 01 AM

File: src/components/interaction-designer/block-types/MobilePrimitives_MessageBlock.vue [hunk 7]
```diff
@@ -13,6 +13,10 @@
         :block="block" />
       <block-name-editor :block="block" />
 
+      <block-output-branching-config
```
Comment (safydy-r) [inline src/components/interaction-designer/block-types/MobilePrimitives_MessageBlock.vue:16]: Got this console error when expand to see configuration panel
Screen Shot 2021-07-22 at 10 56 44 AM

Comment (safydy-r) [reply src/components/interaction-designer/block-types/MobilePrimitives_MessageBlock.vue:16]: Then output branching fire error each time we click on them

File: src/components/interaction-designer/block-types/MobilePrimitives_NumericResponseBlock.vue [hunk 8]
```diff
@@ -24,6 +24,10 @@
         :has-ivr="hasVoiceMode"
         @commitMaxDigitsChange="updateMaxDigits" />
 
+      <block-output-branching-config
```
Comment (safydy-r) [inline src/components/interaction-designer/block-types/MobilePrimitives_NumericResponseBlock.vue:27]: Missing `<hr>` before `<block-output-branching-config>`
(according to design: https://www.figma.com/proto/U2UHUAFgkM0sBdBiinh6ah/DSN-110)

Comment (safydy-r) [reply src/components/interaction-designer/block-types/MobilePrimitives_NumericResponseBlock.vue:27]: For all 03 blocks I think, maybe we just need to insert `<hr>` inside `block-output-branching-config` component directly ?

File: src/components/interaction-designer/Block.vue [hunk 9]
```diff
@@ -135,7 +136,7 @@
             @mouseleave="exitMouseLeave(exit)">
             <span class="block-exit-tag-text align-self-center"
                   v-b-tooltip.hover.top="exit.test">
-              {{ visibleExitTag(key, exit) }}
+              {{exit.name || '(untitled)'}}
```
Comment (safydy-r) [inline src/components/interaction-designer/Block.vue:139]: I think we should localized this `untitled`, if we cannot do it now, let's track it somewhere in a JIRA ticket ;)

File: src/components/interaction-designer/block-editors/BlockOutputBranchingConfig.vue [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+        </button>
+      </div>
+    </div>
+
+    <advanced-exits-builder
+      v-if="isBranchingTypeAdvanced"
+      :block="block" />
+
+    <div class="form-group">
+      <h6>When no valid response/all exit expressions evaluate to false</h6>
```
Comment (safydy-r) [inline src/components/interaction-designer/block-editors/BlockOutputBranchingConfig.vue:63]: Another localization text here @bzabos

File: src/components/interaction-designer/block-editors/BlockOutputBranchingConfig.vue [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+          :id="NoValidResponseHandler.END_CALL"
+          v-model="noValidResponse"
+          class="form-check-input"
+          type="radio"
+          :value="NoValidResponseHandler.END_CALL" />
+
+        <label
+          class="form-check-label"
+          :for="NoValidResponseHandler.END_CALL">
+          End the call/session
```
Comment (safydy-r) [inline src/components/interaction-designer/block-editors/BlockOutputBranchingConfig.vue:76]: Localization !

File: src/components/interaction-designer/block-editors/BlockOutputBranchingConfig.vue [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+          :id="NoValidResponseHandler.CONTINUE_THRU_EXIT"
+          v-model="noValidResponse"
+          class="form-check-input"
+          type="radio"
+          :value="NoValidResponseHandler.CONTINUE_THRU_EXIT" />
+
+        <label
+          class="form-check-label"
+          :for="NoValidResponseHandler.CONTINUE_THRU_EXIT">
+          Continue through Default exit
```
Comment (safydy-r) [inline src/components/interaction-designer/block-editors/BlockOutputBranchingConfig.vue:91]: Localization

File: src/components/interaction-designer/block-editors/BlockOutputBranchingConfig.vue [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+  export default class BlockOutputBranchingConfig extends mixins(Lang) {
+    @Prop() readonly block!: IBlock
+    @Prop() readonly hasExitPerChoice!: boolean
+
+    OutputBranchingType = OutputBranchingType
+    NoValidResponseHandler = NoValidResponseHandler
+
+    mounted(): void {
+      if (isEmpty(this.selectedBranchingType)) {
+        // todo: should this be EXIT_PER_CHOICE for SelectOne block?
```
Comment (safydy-r) [inline src/components/interaction-designer/block-editors/BlockOutputBranchingConfig.vue:144]: is this TODO already answered? If not, who can answer it?

File: src/components/interaction-designer/block-editors/BlockOutputBranchingConfig.vue [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+        case OutputBranchingType.UNIFIED:
+          this.block_convertExitFormationToUnified({blockId})
+          break
+        case OutputBranchingType.EXIT_PER_CHOICE:
+          // handled via change event from SelectOne; felt awkward importing SelectOne store here
+          // should likely make branching types extendable
+          //this.reflowExitsFromChoices({blockId})
+          break
+        case OutputBranchingType.ADVANCED:
+          // todo: restore from cache
```
Comment (safydy-r) [inline src/components/interaction-designer/block-editors/BlockOutputBranchingConfig.vue:171]: Can we track this in a ticket ?

File: src/components/interaction-designer/block-editors/ChoicesBuilder.vue [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+  @Prop() readonly block!: ISelectOneResponseBlock
+
+  draftResource: IResource | null = null
+
+  SupportedContentType = SupportedContentType
+  SupportedMode = SupportedMode
+  findOrGenerateStubbedVariantOn = findOrGenerateStubbedVariantOn
+
+  get choiceKeys(): string[] {
+    // todo: my best guess for ordering is resources ordering? :/
```
Comment (safydy-r) [inline src/components/interaction-designer/block-editors/ChoicesBuilder.vue:78]: this TODO is a bit unclear for me.
who can answer this ?
is it possible to track it in a ticket if it is still relevant ?

File: src/components/interaction-designer/block-editors/ChoicesBuilder.vue [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+    return Object.keys(this.block.config.choices)
+  }
+
+  created(): void {
+    this.generateDraftResource()
+  }
+
+  async generateDraftResource(): Promise<void> {
+    this.draftResource = await this.resource_createWith({ // todo: abstract as createResourceForVanillaChoice()
```
Comment (safydy-r) [inline src/components/interaction-designer/block-editors/ChoicesBuilder.vue:87]: Same in this TODO
is it possible to track it in a ticket if it is still relevant ?

File: src/components/interaction-designer/block-editors/ChoicesBuilder.vue [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+  focusInputElFor(editor?: Vue): void {
+    (editor?.$refs.input as HTMLInputElement).focus()
+  }
+
+  handleExistingResourceVariantChangedFor(resourceId: IResource['uuid'], choiceIndex: number, resource: IResource): void {
+    const isLast = choiceIndex === this.choiceKeys.length - 1
+    const hasEmptyValue = isEmpty(get(resource.values[0], 'value'))
+
+    if (isLast && hasEmptyValue) {
+      // todo: clean up resource, but should we first check for references?
```
Comment (safydy-r) [inline src/components/interaction-designer/block-editors/ChoicesBuilder.vue:119]: same in this TODO
is it possible to track it in a ticket if it is still relevant ?

File: src/components/interaction-designer/block-types/MobilePrimitives_MessageBlock.vue (hunk 7 above)
Comment (safydy-r) [reply src/components/interaction-designer/block-types/MobilePrimitives_MessageBlock.vue:16]: @bzabos , no more an issue, thnx for quick fixing

File: src/components/interaction-designer/block-types/MobilePrimitives_MessageBlock.vue [hunk 18]
```diff
@@ -41,7 +45,7 @@
 import {namespace} from 'vuex-class'
 import {Component, Prop} from 'vue-property-decorator'
 
-import {IFlow, IResource} from '@floip/flow-runner'
+import { IFlow, IResource } from '@floip/flow-runner'
```
[the remark this answers is not in the document]
Comment (safydy-r) [reply src/components/interaction-designer/block-types/MobilePrimitives_MessageBlock.vue:48]: Shouldn't have changed this I think, maybe a formatting issue from IDE?

File: src/fonts/v5icons.svg [hunk 19]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" standalone="no"?>
+<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
+<svg xmlns="http://www.w3.org/2000/svg">
+<metadata>Generated by Fontastic.me</metadata>
+<defs>
+<font id="branches" horiz-adv-x="512">
+<font-face font-family="branches" units-per-em="512" ascent="480" descent="-32"/>
+<missing-glyph horiz-adv-x="512" />
+
+<glyph glyph-name="branching-off" unicode="&#97;" d="M399 263l-281 0 0 190 281 0z m-246 35l211 0 0 120-211 0z m282-240l-142 0 0 123 142 0z m-107 35l72 0 0 53-72 0z m-87 187l35 0 0-74-35 0z m-23-222l-142 0 0 123 142 0z m-107 35l72 0 0 53-72 0z m50 138l194 0 0-35-194 0z m-31-2l35 0 0-61-35 0z m222 0l35 0 0-61-35 0z"/>
```
Comment (safydy-r) [inline src/fonts/v5icons.svg:10]: @bzabos , I'm wondering if we still need these icons ? We're now using new icons from designers, right ?

File: src/lib/custom-icons/index.ts [hunk 20]
```diff
@@ -115,6 +115,45 @@ const minimize: IconDefinition = {
...
+}
+
+const advancedExit: IconDefinition = {
+  prefix: 'fac' as IconPrefix,
+  iconName: 'advanced-exit' as IconName,
+  icon: [
+    32,
+    32,
+    [],
+    'e009',
```
Comment (safydy-r) [inline src/lib/custom-icons/index.ts:137]: should be `e011`, as it's an unique value according to @poojahpatil90

File: src/lib/custom-icons/index.ts [hunk 21]
```diff
@@ -115,6 +115,45 @@ const minimize: IconDefinition = {
...
+}
+
+const oneExitPerChoice: IconDefinition = {
+  prefix: 'fac' as IconPrefix,
+  iconName: 'one-exit-per-choice' as IconName,
+  icon: [
+    32,
+    32,
+    [],
+    'e009',
```
Comment (safydy-r) [inline src/lib/custom-icons/index.ts:150]: and should be `e012` here

File: src/scss/_v5_icons.scss [hunk 22]
```diff
@@ -0,0 +1,49 @@
+// v5icons Font from Fontastic
+// Extra icons should be included into this font files
```
Comment (safydy-r) [inline src/scss/_v5_icons.scss:2]: same questions as here: https://github.com/FLOIP/flow-builder/pull/128/files#r675014808

File: src/store/flow/block-types/MobilePrimitives_SelectManyResponseBlockStore.ts [hunk 23]
```diff
@@ -27,31 +27,28 @@ export const actions: ActionTree<ICustomFlowState, IRootState> = {
...
-    const blankChoicesPromptResource = await dispatch('flow/flow_addBlankResourceForEnabledModesAndLangs', null, {root: true})
-
     const defaultExitProps: Partial<IBlockExit> = {
       uuid: await (new IdGeneratorUuidV4()).generate(),
-      tag: 'Default',
-      label: 'Default',
-      test: '',
+      name: 'Default',
+      default: true,
+      // test: '',
```
Comment (safydy-r) [inline src/store/flow/block-types/MobilePrimitives_SelectManyResponseBlockStore.ts:34]: I'm just curious why we don't need `test` here ?

File: src/store/flow/block-types/MobilePrimitives_SelectOneResponseBlockStore.ts [hunk 24]
```diff
@@ -1,10 +1,16 @@
-import {ActionTree, GetterTree, MutationTree} from 'vuex'
-import {IRootState} from '@/store'
-import {IBlockExit, IBlockExitTestRequired, IResource} from '@floip/flow-runner'
-import {IdGeneratorUuidV4} from '@floip/flow-runner/dist/domain/IdGeneratorUuidV4'
-import {ISelectOneResponseBlock} from '@floip/flow-runner/dist/model/block/ISelectOneResponseBlock'
+import { ActionTree, GetterTree, MutationTree } from 'vuex'
```
[the remark this answers is not in the document]
Comment (safydy-r) [reply src/store/flow/block-types/MobilePrimitives_SelectOneResponseBlockStore.ts:1]: Formatting issue here

File: src/store/flow/block-types/MobilePrimitives_SelectOneResponseBlockStore.ts [hunk 25]
```diff
@@ -1,10 +1,16 @@
...
+import { IRootState } from '@/store'
```
[the remark this answers is not in the document]
Comment (safydy-r) [reply src/store/flow/block-types/MobilePrimitives_SelectOneResponseBlockStore.ts:2]: Formatting issue here

File: src/store/flow/block-types/MobilePrimitives_SelectOneResponseBlockStore.ts [hunk 26]
```diff
@@ -1,10 +1,16 @@
...
+import {
+  IBlock,
+  IBlockExitTestRequired,
+  IBlockExit,
+  IResource, ValidationException,
+} from '@floip/flow-runner'
+import { IdGeneratorUuidV4 } from '@floip/flow-runner/dist/domain/IdGeneratorUuidV4'
```
[the remark this answers is not in the document]
Comment (safydy-r) [reply src/store/flow/block-types/MobilePrimitives_SelectOneResponseBlockStore.ts:8]: Formatting issue here

File: src/store/flow/block-types/MobilePrimitives_SelectOneResponseBlockStore.ts [hunk 27]
```diff
@@ -1,184 +1,124 @@
...
-import {ISelectOneResponseBlock} from '@floip/flow-runner/dist/model/block/ISelectOneResponseBlock'
+import { ActionTree, GetterTree, MutationTree } from 'vuex'
+import { IRootState } from '@/store'
+import {
+  IBlock,
+  IBlockExit,
+  IResource, ValidationException,
+} from '@floip/flow-runner'
+import { IdGeneratorUuidV4 } from '@floip/flow-runner/dist/domain/IdGeneratorUuidV4'
+import { ISelectOneResponseBlock } from '@floip/flow-runner/dist/model/block/ISelectOneResponseBlock'
```
[the remark this answers is not in the document]
Comment (safydy-r) [reply src/store/flow/block-types/MobilePrimitives_SelectOneResponseBlockStore.ts:9]: Formatting issue here

File: src/store/flow/block.ts [hunk 28]
```diff
@@ -78,6 +82,7 @@ export const mutations: MutationTree<IFlowsState> = {
     Vue.delete(findBlockOnActiveFlowWith(blockId, state as unknown as IContext).config!, key)
   },
   block_updateConfigByPath(state, {blockId, path, value}: {blockId: string, path: string, value: object | string}) {
+    // todo: this might still break reactivity
```
Comment (safydy-r) [inline src/store/flow/block.ts:85]: Is the TODO still relevant or just a perso note during implementation?
Might be useful to track it somewhere in JIRA

File: src/store/flow/resource.ts [hunk 29]
```diff
@@ -163,6 +163,15 @@ export function findResourceVariantOverModesOn(
   filter: IResourceDefinitionVariantOverModesFilter,
 ): IResourceValue {
   // b/c we do explicit partial matching on modes
+  // why do we do partial matching on modes?
+
+  // todo: is there a way that we could do an inclusive+additive search here?
```
Comment (safydy-r) [inline src/store/flow/resource.ts:168]: Is the TODO still relevant or just a perso note during implementation?
Might be useful to track it somewhere in JIRA

**Review by safydy-r — APPROVED** [review]
Thanks @bzabos for the hard work here, I can imagine it.
I only went quickly through the code, couldn't have time to think about performance etc :D due to time context.

I put few small concerns there about:
- some localizations
- TODOs
- formatting
- etc

I understand if we cannot address all those TODO now, these are minor.

File: src/components/interaction-designer/block-types/MobilePrimitives_NumericResponseBlock.vue (hunk 8 above)
Comment (bzabos) [reply src/components/interaction-designer/block-types/MobilePrimitives_NumericResponseBlock.vue:27]: I'd prefer to see the `<hr>`s at the block vue file level, it gives most flexibility!

**Pull request merged**